aklinker1/vite-plugin-web-extension
# feat!: v2 rewrite

This is a full restructure of the application's code. Better organized, more extensible, unit testable, etc. The main change is the build process:

#### Old Build Process

```mermaid
graph LR;
  subgraph Primary Vite Build
    A[Build Manifest & All Standalone CSS files & HTML pages]-->B
    subgraph Secondary Vite Builds
      B[Build sandbox HTML]-->C
      C[Build each background.scripts]-->D
      D[Build background.page]-->E
      E[Build background.service_worker]-->F
    end
    F[Build each content_script.js]-->G
  end
  G[Done]
```

#### New Build Process

```mermaid
graph LR;
  subgraph Primary Vite Build
    A[Prepare Manifest]-->K
    subgraph Secondary Vite Builds
      K[Build HTML pages]-->B
      B[Build sandbox HTML]-->C
      C[Build each background.scripts]-->D
      D[Build background.page]-->E
      E[Build background.service_worker]-->
      F[Build each content_script.js]-->J
    end
    J[Build each content_script.css]-->G
    G[Write Manifest]-->H
  end
  H[Done]
```

Essentially, I was combining unrelated inputs into a single vite build, but am no longer. This provides more flexibility around when each individual build can be executed, such as waiting to write the manifest until the other builds have finished so we can user their outputs as the paths in the final manifest.json.

Content script CSS files are now built individually, separate from the HTML files.

### Breaking changes

- No need to use the `generated:` prefix - these files are added to the manifest automatically
- Output format changed from `umd` to `iife` for JS entrypoints
- No more `assets` or `writeAssetsTo` options, use `public` instead
- No `writeManifestTo` option
- Removed `options.serviceWorkerType`
- Require `type: module` in package.json (due to config loading for `web-ext`)
- No more `options.verbose`, use Vite's `--debug` flag instead

### Todo

- [x] #40
- [x] Remove the `assets` option in favor of `public`
- [x] `vite build`
- [x] `vite build --watch`
- [x] Re-integrate `web-ext`
- [x] Manifest validation
- [x] Remove old unused files
- [x] Implement print summary
- [x] Remove `serviceWorkerType`
- [x] Use `ResolvedConfig` instead of `InlineConfig`
   - [x] Calculate directories once
- [x] Remove `options.verbose` in favor of Vite's `--debug` flag
- [x] Restart manifest on change
   - [x] Ensure build context is fully stopped and started back up
- [x] Reload extension manually when build context finishes
- [x] Separate docs for V2
- [x] Make sure `chrome_override_url.*` and `chrome_settings_overrides.homepage` inputs are automatically built along side the popup (see #45)
- [x] Update E2E snapshots
- [x] Write v2 Migration guide
- [x] Include fix for #47 
- [x] HMR
- [x] Make sure https://github.com/aklinker1/vite-plugin-web-extension/commit/1a2788571f92a4581b23efa5002038ffb673ae26 is fixed here too
- [x] Rebase to get latest v1 commits
- [x] Don't crash if the public directory doesn't exist
- [x] Bring over starter kit doc changes to v2 docs: 02a36718b89ba4a3c9b2ace04e5c380d033b6b14
- [x] QA past issues to make sure they are all still working

Future work:

- Update starter kits
- Update websites to only have `v1.vite-plugin-web-extension.aklinker1.io` and `vite-plugin-web-extension.aklinker1.io`
